File: basic_ros_example/src/service_advertiser.cpp
#include <rclcpp/rclcpp.hpp>
#include <basic_ros_example/srv/adder.hpp>

namespace basic_ros_example
{
  class ServiceAdvertiser : public rclcpp::Node {
    public:
      ServiceAdvertiser()
      : rclcpp::Node("service_advertiser")
      {
        service_advertiser_ = create_service<basic_ros_example::srv::Adder>("adder_service",
          std::bind(&ServiceAdvertiser::service_cb, this, std::placeholders::_1, std::placeholders::_2));
      }

    private:
      rclcpp::Service<basic_ros_example::srv::Adder>::SharedPtr service_advertiser_;

      void service_cb(const basic_ros_example::srv::Adder::Request::SharedPtr req,
                           basic_ros_example::srv::Adder::Response::SharedPtr res)
      {
        res->result = req->val1 + req->val2;
      }

  };

}


int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  auto node_instance = std::make_shared<basic_ros_example::ServiceAdvertiser>();
  rclcpp::spin(node_instance);
  rclcpp::shutdown();
  return 0;
}

File: python_package_example/python_package_example/service_advertiser.py
#! /usr/bin/env python3
import rclpy
from rclpy.node import Node
from basic_ros_example.srv import Adder


class ServiceAdvertiser(Node):
    def __init__(self):
        super().__init__('service_advertiser')
        self.srv = self.create_service(srv_name='adder_service', srv_type=Adder, callback=self.srv_callback)

    def srv_callback(self, req: Adder.Request, res: Adder.Response):
        res.result = req.val1 + req.val2
        return res


def main(args=None):
    rclpy.init()
    node_instance = ServiceAdvertiser()
    rclpy.spin(node_instance)

if __name__ == '__main__':
    main()